egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py

```diff
@@ -34,63 +41,138 @@
+        LM rescoring.
+        If False and if `G` is not None, then `num_paths` must be positive
+        and it will use n-best list for LM rescoring.
+      num_paths:
+        Used only if `G` is not None and use_whole_lattice is False.
+        It specify the size of n-best list for LM rescoring.
+      G:
+        The LM. If it is None, no rescoring is used.
+        Otherwise, LM rescoring is used.
+        It supports two types of LM rescoring: n-best list rescoring
+        and whole lattice rescoring.
+        `use_whole_lattice` specifies which type to use.
+
+    Returns:
+      Return the decoding result. See above description for the format of
+      the returned dict.
+    '''
+    device = HLG.device
+    feature = batch['inputs']
+    assert feature.ndim == 3
+    feature = feature.to(device)
+
+    # at entry, feature is [N, T, C]
+    feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+
+    supervisions = batch['supervisions']
+
+    nnet_output, _, _ = model(feature, supervisions)
+    # nnet_output is [N, C, T]
+
+    nnet_output = nnet_output.permute(0, 2, 1)
+    # now nnet_output is [N, T, C]
+
+    supervision_segments = torch.stack(
+        (supervisions['sequence_idx'],
+         (((supervisions['start_frame'] - 1) // 2 - 1) // 2),
+         (((supervisions['num_frames'] - 1) // 2 - 1) // 2)),
+        1).to(torch.int32)
+
+    supervision_segments = torch.clamp(supervision_segments, min=0)
+    indices = torch.argsort(supervision_segments[:, 2], descending=True)
+    supervision_segments = supervision_segments[indices]
+
+    dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+
+    lattices = k2.intersect_dense_pruned(HLG, dense_fsa_vec, 20.0, output_beam_size, 30, 10000)
+
+    if G is None:
+        best_paths = k2.shortest_path(lattices, use_double_scores=True)
+        hyps = get_texts(best_paths, indices)
+        return {'no_rescore': hyps}
+
+    lm_scale_list = [0.7, 0.8, 0.9, 1.0, 1.1, 1.2]
+
+    ans = dict()
+    saved_scores = lattices.scores.clone()
+    for lm_scale in lm_scale_list:
+        lattices.scores = saved_scores.clone()
+        best_paths = decode_with_lm_rescoring(
+            lattices,
+            G,
+            num_paths=num_paths,
+            use_whole_lattice=use_whole_lattice,
+            lm_scale=lm_scale)
+
+        hyps = get_texts(best_paths, indices)
+        key = f'lm_scale_{lm_scale}'
+        ans[key] = hyps
+    return ans
+
+
+@torch.no_grad()
 def decode(dataloader: torch.utils.data.DataLoader, model: AcousticModel,
-           device: Union[str, torch.device], HLG: Fsa, symbols: SymbolTable,
+           HLG: Fsa, symbols: SymbolTable,
            num_paths: int, G: k2.Fsa, use_whole_lattice: bool, output_beam_size: float):
     tot_num_cuts = len(dataloader.dataset.cuts)
     num_cuts = 0
-    results = []  # a list of pair (ref_words, hyp_words)
+    results = defaultdict(list)
```

Comment:
can we have a comment about the structure of `results`?

Reply:
More comments are added.